### PR TITLE
support flashing pca10059 from windows

### PR DESCRIPTION
**Pull request by oflebbe**
New PR since #1767 went wrong , rebased and fixed for current dev.

Still have to look into review comments.

### CODE REVIEW

**Comment by oflebbe** [comment]
@sago35 @aykevl  Regarding https://github.com/tinygo-org/tinygo/pull/1767#discussion_r607408417 : The changes to main.go
are needed because this command line 
`tinygo flash -x -target pca10059 examples/blinky2`
will find the {nrf-dfu} token, and have the fileExt set to "zip" , good.

However in `builder.Build` it will construct the flash command line by replacing the fileExt (aka fileToken i.e. `{zip}`) by the actual file name. But the flash command contains the string `{nrf-dfu}` so it will not replace.
See
https://github.com/tinygo-org/tinygo/blob/164632616496e21f5f11098316c650ab5ad9356c/main.go#L297

**Comment by oflebbe** [comment]
I guess the machine running the test is unstable. Cannot reproduce the failing test (go1.15 with llvm11) locally on macos.

File: targets/pca10059.json [hunk 1]
```diff
@@ -2,5 +2,6 @@
 	"inherits": ["nrf52840"],
 	"build-tags": ["pca10059"],
 	"linkerscript": "targets/pca10059.ld",
-	"flash-command": "nrfutil pkg generate --hw-version 52 --sd-req 0x0 --application {hex} --application-version 1 /tmp/tinygo_$$.zip && nrfutil dfu usb-serial -pkg /tmp/tinygo_$$.zip -p {port} -b 115200 && rm -f /tmp/tinygo_$$.zip"
+	"binary-format": "nrf-dfu",
+	"flash-command": "nrfutil dfu usb-serial -pkg {nrf-dfu} -p {port} -b 115200"
```
Comment (sago35) [inline targets/pca10059.json]: By changing this part to `{zip}`, the changes to `main.go` can be minimized.

```suggestion
	"binary-format": "nrf-dfu",
	"flash-command": "nrfutil dfu usb-serial -pkg {zip} -p {port} -b 115200"
```

Please refer to the `binary-format` and `{bin}` parts of esp32.

https://github.com/tinygo-org/tinygo/blob/release/targets/esp32.json#L18-L19

Comment (oflebbe) [reply targets/pca10059.json:6]: I continue to fail to see how the reference to this example will help me. The {bin} token has an file extension of ".bin".
The nrf-dfu token has a different extension. This is what breaks the logic of the code, IMO. 

Could you please sketch your proposal in more detail ?

Comment (oflebbe) [reply targets/pca10059.json]: Now I got it.

Comment (oflebbe) [reply targets/pca10059.json]: Partially reverted main and squashed. Thanks @sago35 !

**Review by sago35 — APPROVED** [review]
Thank you.
I think it's very good.

File: builder/nrfutil.go [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+	"fmt"
+	"io/ioutil"
+	"os/exec"
+
+	"github.com/tinygo-org/tinygo/compileopts"
+)
+
+// https://infocenter.nordicsemi.com/index.jsp?topic=%2Fug_nrfutil%2FUG%2Fnrfutil%2Fnrfutil_intro.html
+
+func makeDFUFirmareImage(options *compileopts.Options, infile, outfile string) error {
```
Comment (deadprogram) [inline builder/nrfutil.go]: I think you meant `makeDFUFirmwareImage` probably :smile_cat:

Comment (oflebbe) [reply builder/nrfutil.go]: Typo fixed.

**Comment by deadprogram** [comment]
Thanks very much for working on this @oflebbe and thanks for reviewing @sago35 

Now merging.

**Comment by aykevl** [comment]
Looks very good to me!

**Pull request merged**